unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -131,24 +131,31 @@ public boolean contains(final UTF8String substring) {
     }
 
     for (int i = 0; i <= bytes.length - b.length; i++) {
-      // TODO: Avoid copying.
-      if (bytes[i] == b[0] && Arrays.equals(Arrays.copyOfRange(bytes, i, i + b.length), b)) {
+      if (bytes[i] == b[0] && startsWith(substring, i)) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean startsWith(final UTF8String prefix, int offset) {
```

Comment:
Nit: If we use `startsWith(final byte[] prefix, int offset)`, it will may save some calls of `s.getBytes()`. (not sure how is the difference).